detect/main.go

```diff
@@ -46,39 +55,61 @@ func main() {
 		return
 	}
 
-	// Try java
-	if _, ok := detect.BuildPlan[jvm_application_buildpack.JVMApplication]; ok {
-		detect.Logger.Debug("riff Java application")
-		detect.Pass(java.BuildPlanContribution(metadata))
-		return
+	detected := []string{}
+
+	if metadata.Override == "" {
+		// Try java
+		if _, ok := detect.BuildPlan[jvm_application_buildpack.JVMApplication]; ok {
+			detected = append(detected, "java")
+		}
+		// Try node
+		// To be later changed to detecting whether the npm BP voted
+		// see https://github.com/projectriff/riff-buildpack/issues/14
+		if ok, err := node.DetectNode(detect, metadata); err != nil {
+			detect.Logger.Info("Error trying to use node invoker: %s", err.Error())
+			detect.Error(Error_DetectInternalError)
+			return
+		} else if ok {
+			detected = append(detected, "node")
+		}
+		// Try command invoker as last resort
+		if ok, err := command.DetectCommand(detect, metadata); err != nil {
+			detect.Logger.Info("Error trying to use command invoker: %s", err.Error())
+			detect.Error(Error_DetectInternalError)
+			return
+		} else if ok {
+			detected = append(detected, "command")
+		}
+
+		if len(detected) == 0 {
+			detect.Logger.Info("Detected riff function but unable to determine function type.")
+			detect.Error(Error_DetectedNone)
+			return
+		} else if len(detected) > 1 {
```

Comment:
Can be implemented as a pair of guard because of the `return` in each:

```go
if len(detected) == 0 {
	detect.Logger.Info("Detected riff function but unable to determine function type.")
	detect.Error(Error_DetectedNone)
	return
}

if len(detected) > 1 {
	detect.Logger.Info("Detected riff function but ambiguous language detected: %v.", detected)
 	detect.Error(Error_DetectAmbiguity)
 	return
 }
```

Reply:
The return statements are there only to satisfy the tests. `detect.Error()` does `os.Exit()` anyway.
That being said, I personally prefer seeing exclusive cases as `if/else`